src/chatd.cpp

```diff
@@ -519,14 +525,13 @@ void Connection::reset() //immediate disconnect
 
 bool Connection::sendBuf(Buffer&& buf)
 {
-    if (!isOnline())
+    if (!isLoggedIn() && !isConnected())
         return false;
 //WARNING: ws_send_msg_ex() is destructive to the buffer - it applies the websocket mask directly
 //Copy the data to preserve the original
     auto rc = ws_send_msg_ex(mWebSocket, buf.buf(), buf.dataSize(), 1);
     buf.free(); //just in case, as it's content is xor-ed with the websock datamask so it's unusable
-    bool result = (!rc && isOnline());
-    return result;
+    return rc;
```

Comment:
Actually all karere network I/O runs in the karere thread, so the send attempt is done immediately and synchronously. Only monitoring for events and timers runs in the libevent thread.
Even if you want to directly use `rc` you are not using it correctly - it's an int, which if zero means success, if nonzero means failure. You return it as a bool from a function where true means success, false - failure. So you effectively invert the result. As for ignoring the connection state, I agree, that part comes from the time when `isOnline()` checked the libws internal state instead of mState.
For safety, this function should return failure and not even try to send the packet, if we are supposed to be offline. So I would put a check at the start:
`if (!isOnline())
 {
    return;
 }
`